protocol/contracts/nexus/Shard.sol

```diff
@@ -0,0 +1,489 @@
+        totalShardMints[shardId] += amount;
+        _mint(to, shardId, amount, "");
+        /// @dev track event with TransferSingle
+    }
+    
+    /*
+     * @notice Mint shards in batch for allowed minters. 
+     * This is a guarded function which only allowed contracts/addresses can mint.
+     * Function checks if receiving mint address is blacklisted by Relic contract.
+     * @param to The address to mint to
+     * @param shardIds The shard IDs
+     * @param amounts The amount of each shard ID to mint
+     */
+    function mintBatch(
+        address to,
+        uint256[] calldata shardIds,
+        uint256[] calldata amounts
+    ) external isNotBlacklisted(to){
+        uint256 _length = shardIds.length;
+        if (_length != amounts.length) { revert InvalidParamLength(); }
+        uint256 cap;
+        uint256 shardId;
+        uint256 amount;
+        EnumerableSet.UintSet storage minterShards = allowedShardIds[msg.sender];
+        mapping(uint256 => uint256) storage minterShardCaps = allowedShardCaps[msg.sender];
+        uint256 newBalance;
+        for (uint i; i < _length;) {
+            shardId = shardIds[i];
+            if (!minterShards.contains(shardId)) { revert CannotMint(shardId); }
+            amount = amounts[i];
+            cap = minterShardCaps[shardId];
+            newBalance =  mintBalances[msg.sender][shardId] + amount;
+            /// @dev checking only if cap set. 0 (default value) is uncapped
+            if (cap > 0 && cap < newBalance) {
+                revert MintCapExceeded(cap, amount);
+            }
+            mintBalances[msg.sender][shardId] = newBalance;
+            totalShardMints[shardId] += amount;
+            unchecked {
+                ++i;
+            }
+        }
+        _mintBatch(to, shardIds, amounts, "");
+    }
+
+    /*
+     * @notice Burn batch shards. Overriden from base contract. 
+     * Modified to allow Relic contract to burn blacklisted account shards.
+     * @param account The account owning the shard
+     * @param ids The shard IDs
+     * @param values The amounts of each shard to burn
+     */
+    function burnBatch(address account, uint256[] memory ids, uint256[] memory values) public override {
+        // allow relic to burn blacklisted relic shards
+        if (_msgSender() == address(relic)) {
+            _burnBatch(account, ids, values);
+            return;
+        } else if (account != _msgSender() && !isApprovedForAll(account, _msgSender())) {
+            revert ERC1155MissingApprovalForAll(_msgSender(), account);
+        }
+        _burnBatch(account, ids, values);
+    }
+
+    /*
+     * @notice Get shard IDs a minter is allowed to mint
+     * @param minter The minter with perms
+     * @return Shard IDs array
+     */
+    function getMinterAllowedShardIds(address minter) external view returns (uint256[] memory) {
+        return allowedShardIds[minter].values();
+    }
+
+    function nextTokenId() external view returns (uint256) {
```

Comment:
Could just make this public, and then you don't need an identical internal fn (so can remove the _nextTokenId fn)?
```
function nextTokenId() public view returns (uint256) {
    return _currentIndex;
}
```